Spark: Support truncate in FunctionCatalog

kbendick opened the pull request:
This is an offshoot of https://github.com/apache/iceberg/pull/5305 and partially closes https://github.com/apache/iceberg/issues/5349.

Adds a `system.truncate` function that can be used in Spark SQL, as well as can be used as a `FunctionCatalog` function that can be turned into a transform for storage partitioned joins.

This also breaks the definition of `Truncate` out into utility functions inside of `TruncateUtil`. Because different usages validate input at different times, all of the functions in `TruncateUtil` do not validate their input and instead assume that the input is validated by the calling code. This allows for the `Truncate` transforms to validate their width one time (on instantiation), and for the Spark `truncate` function to skip input validation for faster generated code.

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,395 @@
+
+    @Override
+    public String canonicalName() {
+      return "iceberg.truncate[width](tinyint)";
```
Looks like you forgot to actually embed width in the name.

rdblue replied:
Actually, we need to remove `[width]` since the width is passed in.

kbendick replied:
Ok. Should we use `iceberg.truncate(int, tinyint)` or just leave it out entirely?

kbendick replied:
As `canonicalName` is used by Spark to compare if two functions are the same (as function namespacing could be different), leaving the `int` for width isn't necessary.

All functions use `int` for width, so the `canonicalName` functions still correctly identify if a function is equivalent to another.

The actual function's documentation includes `width`, which should be sufficient.

Let me know if you feel differently!

rdblue commented on api/src/main/java/org/apache/iceberg/util/TruncateUtil.java:
```diff
@@ -0,0 +1,80 @@
+    ByteBuffer ret = value.duplicate();
+    ret.limit(Math.min(value.limit(), value.position() + width));
+    return ret;
+  }
```
I think it's a bit confusing to have this as well as `BinaryUtil.truncateBinary`. The javadoc links to `truncateBinary`, but doesn't give any guidance about why there is a nearly identical method here.

I think the motivation for this method is to avoid copying the backing bytes, which seems like a good idea. Instead of adding this here, I would like to see it move to `BinaryUtil` and given a name that distinguishes it from the behavior of `truncateBinary`. Maybe `truncateBinaryUnsafe` or something.

kbendick replied:
I had thought about using the `truncateBinaryUnsafe` idea (even with that name). I'll give that a go.

kbendick replied:
Moved over to `BinaryUtil#truncateBinaryUnsafe`.

rdblue commented on api/src/main/java/org/apache/iceberg/util/TruncateUtil.java:
```diff
@@ -0,0 +1,80 @@
+    return value - (((value % width) + width) % width);
+  }
+
+  public static long truncateLong(int width, long value) {
```
Shouldn't this `width` be a long?

rdblue commented on core/src/test/java/org/apache/iceberg/util/TestTruncateUtil.java:
```diff
@@ -0,0 +1,99 @@
+        TruncateUtil.truncateDecimal(BigInteger.valueOf(width), BigDecimal.valueOf(1234L, 2)));
+
+    int scale = 2;
+    BigDecimal input = BigDecimal.valueOf(1234L, scale);
```
Why use a local variable for `scale` here, but not above? It isn't referenced anywhere because the expected value is `input.scale()`

kbendick replied:
Removed.

rdblue commented on core/src/test/java/org/apache/iceberg/util/TestTruncateUtil.java:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  @Test
+  public void testTruncateByte() {
```
I don't find these test cases very useful. They're difficult to read because you're using `MAX_VALUE` to pass both width and value, so it isn't clear which one is which (`truncateByte(10, 244)` is much more clear).

I also think there are already tests for most of these cases, so I'd probably remove these and make sure the other paths are covered.

kbendick replied:
Will remove. The pathway is already covered by `TestSparkTruncateFunction`.

kbendick replied:
Removed all of these and left just the test that shows it may or may not throw an exception for invalid input.

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,395 @@
+import org.apache.spark.unsafe.types.UTF8String;
+
+/**
+ * A function for use in SQL that truncates input to a given width according to the rules of the
```
I think I already pointed out that "for use in SQL" isn't quite accurate and is not helping the reader understand. Can you remove it?

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,395 @@
+/**
+ * A function for use in SQL that truncates input to a given width according to the rules of the
+ * existing Truncate partition transform, e.g. {@code SELECT system.truncate(1, 'abc')} will return
+ * the String 'a'.
```
I think this should be `A Spark function implementation for the Iceberg truncate transform.`

kbendick replied:
I like that as well.

Updated to that phrasing, while keeping the example SELECT statement code. I can remove that as well, though I think it provides benefit.

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,395 @@
+          DataTypes.ByteType,
+          DataTypes.ShortType,
+          DataTypes.IntegerType,
+          DataTypes.LongType,
```
What about `DecimalType`?

kbendick replied:
This is just the atomic types, as `DataTypes.DecimalType` cannot be used (a precision and scale is needed).

In the validation for the truncation field, we include DecimalType the following way.

```java
  private static void validateTruncationFieldType(DataType dt) {
    if (truncateableAtomicTypes.stream().noneMatch(type -> type.sameType(dt))
        && !(dt instanceof DecimalType)) {
```

Let me see if `DataTypes.createDecimalType()` will work and ignore the precision / scale so they can all be kept in one set.

kbendick replied:
No, `DataTypes.createDecimalType()` doesn't work as the type generated from that will compare with `precision` and `scale` to the input type to be truncated.

I think we can just remove this function, given that the actual call to bind that this function is used in does type validation already when choosing a specific input type (as follows):
```java
    DataType toTruncateDataType = toTruncateField.dataType();
    if (toTruncateDataType instanceof ByteType) {
      return new TruncateTinyInt();
    } else if (toTruncateDataType instanceof ShortType) {
      return new TruncateSmallInt();
    } else if (toTruncateDataType instanceof IntegerType) {
      return new TruncateInt();
    } else if (toTruncateDataType instanceof LongType) {
      return new TruncateBigInt();
    } else if (toTruncateDataType instanceof DecimalType) {
      return new TruncateDecimal(
          ((DecimalType) toTruncateDataType).precision(),
          ((DecimalType) toTruncateDataType).scale());
    } else if (toTruncateDataType instanceof StringType
        || toTruncateDataType instanceof VarcharType
        || toTruncateDataType instanceof CharType) {
      return new TruncateString();
    } else if (toTruncateDataType instanceof BinaryType) {
      return new TruncateBinary();
    } else {
      throw new UnsupportedOperationException("Cannot truncate type: " + toTruncateDataType);
    }
```

rdblue replied:
I guess just the name is misleading. I wrote this before I saw how it was used. This isn't a complete list of truncateable types, it's just the ones that can be checked using `sameType`.

A better name would help. Also, `static final` constants should use names that are ALL_UPPER_CASE.

kbendick replied:
We might need to normalize the Varchar / Char types, but otherwise I'll get rid of the function and just leave what is here.

kbendick replied:
Got rid of the function and am now depending on this input block.

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,395 @@
+    if (truncateableAtomicTypes.stream().noneMatch(type -> type.sameType(dt))
+        && !(dt instanceof DecimalType)) {
+      String expectedTypes =
+          "[ByteType, ShortType, IntegerType, LongType, StringType, BinaryType, DecimalType]";
```
We generally avoid local variables to hold static strings, especially when the string is going to be passed to `format`.

Can you embed this in the format string instead?

kbendick replied:
Sure thing.

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,395 @@
+    if (inputType.fields().length != 2) {
+      throw new UnsupportedOperationException(
+          String.format(
+              "Invalid input type. Expected 2 fields but found %s", inputType.fields().length));
```
It's unlikely that the caller knows what you mean by "input type". While Spark passes `inputType` as a single struct here, the caller's view of the function is `truncate(6, col)`. Instead, I think this should be `Cannot bind truncate: wrong number of inputs (expected width and value)`

kbendick replied:
Updated.

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,395 @@
+              "Invalid input type. Expected 2 fields but found %s", inputType.fields().length));
+    }
+
+    StructField widthField = inputType.apply(0);
```
Minor: I'd prefer `inputType.fields()[0]` rather than using `apply` because it isn't clear what `apply` does. It's named `apply` so that Scala syntactic sugar works.

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,395 @@
+          ((DecimalType) toTruncateDataType).scale());
+    } else if (toTruncateDataType instanceof StringType
+        || toTruncateDataType instanceof VarcharType
+        || toTruncateDataType instanceof CharType) {
```
Varchar and Char aren't allowed by `validateTruncationFieldType`. Do these `instanceof` checks work?

kbendick replied:
Yes these `instanceof` checks work the same way that that `instanceof DecimalType` works (because `VarcharType` and `CharType` both have a size / length field so they cannot be compared directly).

I can update `validateTruncationFieldType`. However, I think that `validateTruncationFieldType` is somewhat unnecesssary given that this whole block basically does the same thing. Let me try moving it around a bit.

rdblue replied:
Yeah, I don't think that method is necessary. Let's remove it.

kbendick replied:
Removed the method.

I'm going to remove `varchartype` and `chartype` as Spark prints a warning saying that it casts these two aliases for `UTF8String` to string already and that neither type can be used for functions etc.

kbendick replied:
Yep. Removed the check for `VarcharType` and `CharType` as they're not necessary.

There are unit tests to verify this behavior and it also matches the Spark docs.

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,395 @@
+  @Override
+  public String description() {
+    return "Truncate - The Iceberg truncate function used for truncate partition transformations.\n"
+        + "\tCalled with the truncation width as the first argument: e.g. system.truncate(width, col)";
```
I thought this previously used `name()`? I think it should use `name()` to ensure the name is consistent. I would also make this more like docs, where the call is documented:

```
truncate(width, col) - Call Iceberg's truncate transform
  width :: width for truncation, e.g. truncate(10, 255) -> 250 (must be an integer)
  col :: column to truncate (must be an integer, decimal, string, or binary)
```

kbendick replied:
Good call. That looks much cleaner. Updated to use `name()` and changed to this documentation 👍

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,395 @@
+    }
+  }
+
+  private static Integer readAndValidateWidth(InternalRow input) {
```
The logic here is odd. Why call `isNullAt` to box the value only to check whether it is null and throw an exception? I'd expect something like this:

```java
  if (input.isNullAt(0)) {
    throw IllegalArgumentException("Invalid truncation width: null");
  }

  return width;
```

I left out checking whether width is negative. I don't think we care.

kbendick replied:
Removed the boxing.

For reference, the magic functions that take in only primitives will return `null` if the input width (an int) is `null`. But I think this is fine as `produceResult` isn't really called anyway.

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,395 @@
+      Integer width = readAndValidateWidth(input);
+
+      Byte toTruncate = !input.isNullAt(1) ? input.getByte(1) : null;
+      return toTruncate != null ? invoke(width, toTruncate) : null;
```
This does an identical null check twice. I think this should be:

```java
return `input.isNullAt(1) ? null : invoke(width, input.getByte(1));`
```

kbendick replied:
Changed to this.

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,395 @@
+
+  public static class TruncateString extends TruncateBase<UTF8String> {
+    // magic function for usage with codegen
+    // TODO - This can likely be made more efficient, but using existing definition for now.
```
How would you make this more efficient?

kbendick replied:
I think we could use the built-in `substring` method or `substringSQL`, which both seem to take into account the logic of `numChars()` method.

`substring`  specifically mentions that it works by code-point: https://github.com/apache/spark/blob/d77bc7027675b9823bcee96589eb0dddee3cfaa0/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L264-L294

kbendick replied:
Javadoc for `UTF8String::substring`

```
  /**
   * Returns a substring of this.
   * @param start the position of first code point
   * @param until the position after last code point, exclusive.
   */
```

kbendick replied:
Tried it with `substring` and it passes all of the tests, including utf-8 unicode character aware truncation tests.

Updated to use that method and reverting the other previous changes.

rdblue commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java:
```diff
@@ -0,0 +1,377 @@
+
+  @Test
+  public void testTruncateInt() {
+    Assert.assertEquals(0, scalarSql("SELECT system.truncate(10, %d)", 0));
```
Why are the ints put into SQL using a format string? None of the other tests do that.

kbendick replied:
Hold over from a previous version. Updated!

rdblue commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java:
```diff
@@ -0,0 +1,377 @@
+    // decimal truncation works by applying the decimal scale to the width: ie 10 scale 2 = 0.10
+    Assert.assertEquals(
+        new BigDecimal("12.30"),
+        scalarSql("SELECT system.truncate(10, CAST(%f as DECIMAL(9, 2)))", 12.34));
```
When working with decimals, you want to prefer strings. Formatting a decimal may give a different value and it looks like this assumes the value will be what was given as a float.

kbendick replied:
Good point. Updated to use format strings with string values for the decimal test cases.

rdblue commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java:
```diff
@@ -0,0 +1,377 @@
+        new BigDecimal("12345.3480"),
+        scalarSql("SELECT system.truncate(10, 12345.3482)"));
+
+    int scale = 4;
```
Why have a variable for this if it isn't used everywhere?

kbendick replied:
Removed.

rdblue commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java:
```diff
@@ -0,0 +1,377 @@
+  }
+
+  @Test
+  public void testInvalidTypesCannotBeUsedForWidth() {
```
Is there a test that a short or byte can be used as a width?

kbendick replied:
Yeah I have tests for that already. But I'll make it more explicit.

kbendick replied:
Added `testWidthAcceptsShortAndByte`

rdblue commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java:
```diff
@@ -0,0 +1,377 @@
+    Assert.assertEquals(
+        "Null input should return null",
+        null,
+        scalarSql("SELECT system.truncate(2, CAST(null AS int))"));
```
Does this call `produceResult` or `invoke`?

kbendick replied:
This calls `invoke`. For `invoke`, if any primitive input is `null`, `null` is returned automatically when using the magic functions -- for non-primitives like UTF8String, we still need to check.

Here's the query plan to show it:
```
    == Parsed Logical Plan ==
    'Project [unresolvedalias('system.truncate(2, cast(null as bigint)), None)]
    +- OneRowRelation
    == Analyzed Logical Plan ==
    staticinvoke(class org.apache.iceberg.spark.functions.TruncateFunction$TruncateBigInt, LongType, invoke, 2, cast(NULL as bigint), IntegerType, LongType, false, true, true): bigint
    Project [staticinvoke(class org.apache.iceberg.spark.functions.TruncateFunction$TruncateBigInt, LongType, invoke, 2, cast(null as bigint), IntegerType, LongType, false, true, true) AS staticinvoke(class org.apache.iceberg.spark.functions.TruncateFunction$TruncateBigInt, LongType, invoke, 2, cast(NULL as bigint), IntegerType, LongType, false, true, true)#55L]
    +- OneRowRelation
    == Optimized Logical Plan ==
    Project [null AS staticinvoke(class org.apache.iceberg.spark.functions.TruncateFunction$TruncateBigInt, LongType, invoke, 2, cast(NULL as bigint), IntegerType, LongType, false, true, true)#55L]
    +- OneRowRelation
    == Physical Plan ==
    *(1) Project [null AS staticinvoke(class org.apache.iceberg.spark.functions.TruncateFunction$TruncateBigInt, LongType, invoke, 2, cast(NULL as bigint), IntegerType, LongType, false, true, true)#55L]
    +- *(1) Scan OneRowRelation[]
```

rdblue commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java:
```diff
@@ -0,0 +1,377 @@
+  }
+
+  @Test
+  public void testMagicFunctionsResolveForTinyIntAndSmallIntWidths() {
```
Can you put this by the negative tests?

kbendick replied:
Done. Moved all of the `width` related tests to be in one block towards the end -- or more specifically after the test suites for each individual type to be truncated.

kbendick commented:
cc @rdblue PTAL when you have time. I've addressed all of your feedback. Thanks!

kbendick commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,348 @@
+        return null;
+      }
+
+      return value.substring(0, width);
```
The built-in `substring` function of UTF8String is almost certainly faster than converting to `ByteBuffer` and then `CharSequence` so we can use the existing `StringBuilder` based `UnicodeUtil` method for code-point based truncation, so using that.

There's an existing function in `UTF8String`, called `toTitleCaseSlow` which uses our StringBuilder based approach, so I assume that the methods which rely on the bytes themselves and are code-point aware are faster than doing all of those conversions.

aokolnychyi replied:
I'd be up to using UTF8String directly. Do we have enough tests to verify the bevahior of the Spark function will match the behavior of Iceberg transform?

rdblue replied:
This is a good point. I added a comment here to follow up and add all of our unicode truncation tests: https://github.com/apache/iceberg/pull/5431/files#r943717037

kbendick replied:
I agree with that.

But I took a look and I actually couldn't find any test cases for `UnicodeUtil.truncateString` directly.

I looked up the tests for `TestTruncate`, but `TestTruncate#testTruncateString` has _very few_ tests - all of which are covered here.

https://github.com/apache/iceberg/blob/5a15efc070ab59eeda6343998aa065c0c9892c5c/api/src/test/java/org/apache/iceberg/transforms/TestTruncate.java#L70-L77

I was able to pull more test cases from `TestMetricsTruncation`, and we now cover 2-byte, 3-byte, and 4-byte unicode character truncation (as well as mixed byte alignment).

So I'd say we definitely have enough tests for string truncation now (at least compared to elsewhere).

kbendick replied:
If we want to create a followup task to add more of these tests _back_ to `TestTruncate` or to a `UnicodeUtil` test suite, let me know as I could easily handle that.

At present, this code seems well tested though for using the built-in `UTF8String` tests.

kbendick commented:
cc @nastra and @aokolnychyi as well as you reviewed the FunctionCatalog PR itself 🙂

kbendick commented on api/src/main/java/org/apache/iceberg/transforms/Truncate.java:
```diff
@@ -252,7 +254,7 @@ public CharSequence apply(CharSequence value) {
       }
 
-      return UnicodeUtil.truncateString(value, length);
+      return UnicodeUtil.truncateStringUnsafe(length, value);
```
This change keeps the same logic but avoids checking the `width` in a precondition for every string.

The width is already validated in the constructor and this is a hot code path.

rdblue replied:
Why are the arguments transposed? I think the new method should use the same argument order that the other `UnicodeUtil` method does.

kbendick replied:
The arguments were transposed to match the other ones from `TruncateUtil`. But I'll update it as it is somewhat odd to be transposed within `UnicodeUtil`.

kbendick replied:
This wound up being removed anyway. In a separate PR, I'll update this `Transform` path to not validate the input width on every call but it's out of scope for this PR.

kbendick replied:
I opened https://github.com/apache/iceberg/pull/5496 to skip validating the input `length` for CharSequence Truncate on every record.

rdblue commented on api/src/main/java/org/apache/iceberg/util/BinaryUtil.java:
```diff
@@ -50,6 +50,22 @@ public static ByteBuffer truncateBinary(ByteBuffer input, int length) {
+  /**
+   * Truncates the input byte buffer to the given length.
+   *
+   * <p>This skips input validation, and avoids copying the backing bytes if possible for efficiency
```
Why does this say "if possible"? This always returns a duplicate of the original buffer.

The distinction between this and `truncateBinary` is that this doesn't copy the data. You can remove any wording about efficiency or input validation. The buffer handles validation and will throw IllegalArgumentException if needed.

kbendick replied:
The main part remaining about skipping input validation is that `width` isn't validated to be non-negative. I can leave that out though if desired.

rdblue replied:
Yep. The buffer handles validation.

rdblue commented on api/src/main/java/org/apache/iceberg/util/UnicodeUtil.java:
```diff
@@ -40,6 +40,17 @@ public static boolean isCharHighSurrogate(char ch) {
+   * and the number of unicode characters in the truncated charSequence is lesser than or equal to
+   * length.
+   *
+   * <p>Note that this function does not validate the input
```
Because this uses `UTF8String.substring`, this new method is not called. Can you please revert the changes here and in `Truncate`?

kbendick replied:
Reverted. I might open another PR with these changes as right now the `truncateString` call in the truncate transform validates the input length for every input which is unnecessary.

But it's out of scope for this PR. Updated.

kbendick replied:
Opened https://github.com/apache/iceberg/pull/5496 to avoid validating the input on every record in the CharSequence Truncate transform.

But it's out of scope for this PR so removed from here 👍

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,348 @@
+
+  // Not identical behavior to magic methods used in codegen,
+  // This function throws on a `null` width, to avoid boxing.
+  // The magic method in Spark's codegen will return `null` for a `null` width.
```
What's the value of different behavior? We could also return `null` rather than throwing an exception. That seems more reasonable to me. Otherwise, you get different behavior depending on how Spark decided to call this function.

kbendick replied:
Updated and removed.

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,348 @@
+    @Override
+    public byte[] produceResult(InternalRow input) {
+      int width = width(input);
+      return input.isNullAt(1) ? null : invoke(width, input.getBinary(1));
```
Let's get rid of `width` and update these to `input.isNullAt(0) || input.isNullAt(1) ? null : invoke(input.getInt(0), input.getBinary(1))`

rdblue replied:
Actually, I made this comment in the wrong place. That should be for the primitive `invoke` functions. Here, we can pass `input.getBinary(1)` through to `invoke` without checking it because invoke checks for null.

kbendick replied:
Will skip the null checks for the ones that `invoke` checks for null.

kbendick replied:
Actually added the null check back in, as certain specialized accessors might throw instead of returning null if the underlying value is null.

Figured it was safer to just keep the implementation consistent.

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,348 @@
+      throw new UnsupportedOperationException(
+          String.format(
+              "Invalid input type to truncate. Expected one of [ByteType, ShortType, IntegerType, LongType, "
+                  + "DecimalType, StringType, BinaryType], but found %s",
```
These names are internal Spark names for types. I think we should use the SQL names instead:

```
Invalid input types: truncate(width=%s, col=%s), expected col to be tinyint, shortint, int, bigint, decimal, string, or binary
```

kbendick replied:
Agree on updating the data types to use the SQL types. But Spark adds some to this error message. Let me show you an example.

When passing `(int, interval year to month)` types, I get the following message now.

```
 "Function 'truncate' cannot process input: (int, interval year to month): Invalid input type to truncate. Expected one of [ByteType, ShortType, IntegerType, LongType, DecimalType, StringType, BinaryType], but found YearMonthIntervalType(0,1)"
```

So I would say that we don't need to repeat the name of the function necessarily as Spark adds it. I'd suggest using 
`expected truncation col to be tinyint, shortint, int, bigint, decimal, string, or binary`, so that the error message will wind up being:

`" Function 'truncate' cannot process input: (int, interval year to month): expected truncation col to be tinyint, shortint, int, bigint, decimal, string, or binary"`

kbendick replied:
TLDR - Spark prepends `Function 'truncate' cannot process input: (int, interval year to month): ` to the error message from the analysis exception.

rdblue replied:
Okay, then we don't need all the context. In that case, just change it to `Expected tinyint, shortint, int, long, decimal, string, or binary: interval year to month `

kbendick replied:
I updated the test error messages to show the full message that spark adds (which is very similar to the one you proposed).

kbendick commented on api/src/main/java/org/apache/iceberg/util/BinaryUtil.java:
```diff
@@ -50,6 +50,22 @@ public static ByteBuffer truncateBinary(ByteBuffer input, int length) {
+   * @param width The non-negative length to truncate input to
+   * @param value The ByteBuffer to be truncated
+   */
+  public static ByteBuffer truncateBinaryUnsafe(int width, ByteBuffer value) {
```
Sure. They were ordered this way to match the spark functions ordering and the ordering of the arguments of the `TruncateUtil`, but I agree it's confusing to have them in a different order within the same file.

aokolnychyi commented:
I'd love to take a look as well today.

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,340 @@
+
+  @Override
+  public BoundFunction bind(StructType inputType) {
+    if (inputType.fields().length != 2) {
```
nit: just `inputType.length()` or `inputType.size()`?

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,340 @@
+        && !DataTypes.ShortType.sameType(widthType)
+        && !DataTypes.ByteType.sameType(widthType)) {
+      throw new UnsupportedOperationException(
+          "Expected truncation width to be tinyint, shortint or int");
```
nit: I'd probably either use `Cannot bind truncate:` prefix in all error messages related to binding or drop it everywhere for consistency.

rdblue replied:
I think this is handled by Spark. We had a short discussion about error messages in a different thread and if we throw `UnsupportedOperationException` then Spark fills in what failed and the input types.

kbendick replied:
Yeah. I'll remove it from the length check.

Let me check what Spark adds for length itself. We might not even need a check at all (just a unit test) for length.

kbendick replied:
We did need a length check, but the error message didnd't need to include the `Cannot bind truncate:`.

I simplified it and added tests.

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,340 @@
+
+    StructField widthField = inputType.fields()[0];
+    StructField toTruncateField = inputType.fields()[1];
+    validateTruncationWidthType(widthField.dataType());
```
Do we have to validate width nullability?

kbendick replied:
Let me see if the logic added for storage-partitioned joins creates a`bucket` function that uses a non-null integer type.

I know this is `truncate`, but that can inform the decision here.

kbendick replied:
So it looks like the `bucket` transform in Spark creartes a non-null integer field. So we can check that here too and it seems to reject correctly. Updating!

https://github.com/apache/spark/blob/126870eddc57fff7408942897dd59a2c19c91c4a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/V2ExpressionUtils.scala#L90-L100

kbendick replied:
I was able to get failure when the `width` field is explicitly non-null. I added two tests for it as well.

I'm not sure if we should remove the checks for null in `produceResult` as Spark isn't great at tracking null vs not null. But it seems that if the schema is `nullable` that function binding will fail, so I personally think that's sufficient to remove the check for `result.isNullAt(WIDTH_ORDINAL)`. WDYT?

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,340 @@
+    }
+
+    StructField widthField = inputType.fields()[0];
+    StructField toTruncateField = inputType.fields()[1];
```
nit: Do we call it `value` in most other places?

kbendick replied:
Updated

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,340 @@
+
+    StructField widthField = inputType.fields()[0];
+    StructField toTruncateField = inputType.fields()[1];
+    validateTruncationWidthType(widthField.dataType());
```
What do you think about adding an empty line before this call to separate the block that instantiates the fields from the validation?

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,340 @@
+
+    @Override
+    public Byte produceResult(InternalRow input) {
+      return input.isNullAt(0) || input.isNullAt(1)
```
What about defining constants for 0 and 1 with some reasonable names to indicate what those are?

```
private static final int WIDTH_ORDINAL = 0;
private static final int VALUE_ORDINAL = 1;
```

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,340 @@
+
+    @Override
+    public Short produceResult(InternalRow input) {
+      return input.isNullAt(0) || input.isNullAt(1)
```
Do we actually allow nullable width?

rdblue replied:
I think it is better if we do because Spark doesn't do a very good job of tracking whether a value is nullable. Lots of values are apparently nullable even though they aren't really.

aokolnychyi replied:
I know but I think this will be a constant, right? I am talking about width only.

kbendick replied:
The reason for this is that Spark will return `null` for the code-generated code with `null` input. And we don't have much control over that.

I can try setting the input schema that's used for binding and see if requiring that to be non-null will be tracked and throw on a `null` input. But in that case we'd require the user to cast the `width` column (if one is used) to a non-null integer.

We've already taken the standpoint that bad input on the width will lead to undefined results somewhat (in that we aren't testing the width). Technically, the width can be variable and allow for a different value for each column - though in practice we don't expect to use it this way.

This two argument set up is also how the `bucket` v2 function is expected to work.

Let me check if Spark will respect declaring the input schema as non-null.

aokolnychyi replied:
Okay, I took a look at `ScalarFunction`.

```
 * <h2> Handling of nullable primitive arguments </h2>
 * The handling of null primitive arguments is different between the magic method approach and
 * the {@link #produceResult} approach. With the former, whenever any of the method arguments meet
 * the following conditions:
 * <ol>
 *   <li>the argument is of primitive type</li>
 *   <li>the argument is nullable</li>
 *   <li>the value of the argument is null</li>
 * </ol>
 * Spark will return null directly instead of calling the magic method. On the other hand, Spark
 * will pass null primitive arguments to {@link #produceResult} and it is user's responsibility to
 * handle them in the function implementation.
```

Then the existing approach makes more sense to me. At least, it is in line with `invoke`. Let's keep it as-is.

kbendick replied:
I was able to get failure when the width field is explicitly non-null. I added two tests for it as well.

But Spark isn't great at tracking null vs not null.

So I'm ok either way. It's also arguably a bit harder on the user if they're going to pull `width` from a column, as Spark tends to make things nullable pretty often. So I'm good either way. I can revert the changes I made.

kbendick replied:
I reverted the check for `width` field nullability and the associated tests.

I think this is the right choice as Spark is notoriously not great at tracking whether a field is non-null or not. But here's the commit that has the check and tests in case you're curious: https://github.com/apache/iceberg/pull/5431/commits/ad5343facb561750f24e59e8cdba4da882b0313c

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,340 @@
+
+    @Override
+    public Long produceResult(InternalRow input) {
+      return input.isNullAt(0) || input.isNullAt(1)
```
I wonder whether explicit if statements would be more readable given that we have to split the ternary operation onto multiple lines.

kbendick replied:
Good call. I'm not a fan of the ternary style that spotless adds.

rdblue commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java:
```diff
@@ -0,0 +1,436 @@
+    Assert.assertEquals(
+        "Should handle three-byte UTF-8 characters mixed with two byte utf-8 characters",
+        "测试ra",
+        scalarSql("SELECT system.truncate(4, '测试raul试测')"));
```
Because we are using Spark's truncate function, I think this needs to run through all of the test cases for Iceberg's `UnicodeUtil.truncateString` and make sure we get the expected output.

kbendick replied:
I agree with that.

But I took a look and I actually couldn't find any test cases for `UnicodeUtil.truncateString` directly.

I looked up the tests for `TestTruncate`, but `TestTruncate#testTruncateString` has _very few_ tests - all of which are covered here.

https://github.com/apache/iceberg/blob/5a15efc070ab59eeda6343998aa065c0c9892c5c/api/src/test/java/org/apache/iceberg/transforms/TestTruncate.java#L70-L77

kbendick replied:
Let me see if I can find more test cases somewhere 🙂

kbendick replied:
`TestMetricsTruncation` has more test cases, though they're specifically for `testTruncateMin` and `testTruncateMax`.

But I can reuse those as they do cover more situations: https://github.com/apache/iceberg/blob/5a15efc070ab59eeda6343998aa065c0c9892c5c/core/src/test/java/org/apache/iceberg/TestMetricsTruncation.java#L132-L264

kbendick replied:
I copied over the tests from `TestmetricsTruncation` to ensure that 2-byte, 3-byte, and 4-byt UTF-8 codepoints were all covered.

So we should be fully covered in terms of testing truncate String (at least as covered as we are elsewhere).

rdblue replied:
Looks good.

kbendick commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java:
```diff
@@ -0,0 +1,340 @@
+public class TruncateFunction implements UnboundFunction {
+
+  private static void validateTruncationWidthType(DataType widthType) {
+    if (!DataTypes.IntegerType.sameType(widthType)
```
Good call. I had a similar setup before for the truncation type, but it required that I have `DecimalType` so the pattern became less helpful.

This works well though! Updated.

kbendick commented:
@aokolnychyi @rdblue I addressed all of your feedback.

Please take a look when you get a chance. Also, on the subject of nullability for the `width` column, I was able to make it work but reverted it based on this comment thread: https://github.com/apache/iceberg/pull/5431#discussion_r943914954

Knowing that Spark isn't the best at keeping track of nullability, I think this is better and adheres to the contract laid out in `ScalarFunction` as quoted by @aokolnychyi.

But take a look at the commit that added nullability-checking on the width field if you'd like: https://github.com/apache/iceberg/pull/5431/commits/ad5343facb561750f24e59e8cdba4da882b0313c

rdblue commented:
Looks good to me. @aokolnychyi, do you want to take another look?

aokolnychyi commented:
@rdblue, let me take a quick now.

aokolnychyi reviewed (approved):
Looks great!

aokolnychyi commented:
Thanks, @kbendick! Could you cherry-pick this to 3.2?

kbendick commented:
Here's the PR for bucket. All the feedback from this PR was more or less applied there as well - https://github.com/apache/iceberg/pull/5513

kbendick commented:
> Thanks, @kbendick! Could you cherry-pick this to 3.2?

Sure thing!
